### ISSUE TITLE
Logout functionality fails for sub organizations when using IS as federated IDP in APIM 

### ISSUE THREAD

**Issue by HiranyaKavishani** [wso2/product-is#21927]
**Describe the issue:**
We have configured IS 7.1-M5 as a federated Identity Provider (IDP) for APIM. However, we have observed that the logout functionality is not working correctly for users in sub-organizations. It appears that the logout call is not being properly propagated from the root organization to the sub-organizations, causing inconsistencies in session termination.

**How to reproduce:**
1. Setup IS 7.1-M5 as federated IDP for APIM 4.4.0 or any other IS
2. Create an App using `Traditional Web Application` option from super org
3. Create a sub org, switch to that org and create a new user 
4. Add the necessary role to user in order to access the APP
5. Login to APP using that sub org user and try to logout

**Expected behavior:**
logout function for sub org users should be worked properly.

**Environment information**
 - Product Version: IS 7.x

---

### Optional Fields
**Related issues:**

**Suggested labels:**

**Comment by chamilaadhi** [comment]
When reproducing please check with @ShanChathusanda93 since he has context on this

**Comment by Yasasr1** [comment]
I did some initial analysis on this issue. As mentioned above I was able to reproduce the issue with two IS instances.

![issue](https://github.com/user-attachments/assets/79c1e74e-8b73-40d9-a3b2-c97f610e5d53)

The application sends the logout request A to IS1 as expected. This request includes the OPBS cookie which is used to identify the session. With the cookie, IS1 identify the session successfully and invokes the federated IDP to initiate a logout request to IS2 [1] [2]

This logout request to IS2 does not contain the OPBS cookie. Due to this the check [3] fails and IS2 is not calling the organization SSO authenticator to send a logout request to the sub-organization. So request C is never sent

[1] - https://github.com/wso2-extensions/identity-inbound-auth-oauth/blob/v7.0.203/components/org.wso2.carbon.identity.oidc.session/src/main/java/org/wso2/carbon/identity/oidc/session/servlet/OIDCLogoutServlet.java#L739
[2] - https://github.com/wso2/carbon-identity-framework/blob/v7.7.52/components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/request/impl/DefaultLogoutRequestHandler.java#L298
[3] - https://github.com/wso2-extensions/identity-inbound-auth-oauth/blob/v7.0.203/components/org.wso2.carbon.identity.oidc.session/src/main/java/org/wso2/carbon/identity/oidc/session/servlet/OIDCLogoutServlet.java#L167

**Comment by chamilaadhi** [comment]
Hi @AnuradhaSK 
Since this is needed for upcoming APIM organization visibility feature, hope you could prioritize this issue

**Comment by AnuradhaSK** [comment]
> Hi [@AnuradhaSK](https://github.com/AnuradhaSK) Since this is needed for upcoming APIM organization visibility feature, hope you could prioritize this issue

Yes this is in the priority list and we will attend to fix this on 28th Jan

**Comment by SujanSanjula96** [comment]
After further analysis, we observed that when there is no OPBS cookie in OIDC logout request, there is a logic to handle the missing session state gracefully [1]. But it doesn't send a logout request to framework which avoid sending OIDC logout from Organization SSO authenticator. We should handle the OIDC logout to properly redirect to framework logout.


[1] - https://github.com/wso2-extensions/identity-inbound-auth-oauth/blob/0c8ed64bac23ddf5d29b3af3cd411eda1683c16a/components/org.wso2.carbon.identity.oidc.session/src/main/java/org/wso2/carbon/identity/oidc/session/servlet/OIDCLogoutServlet.java#L173

cc: @DMHP

**Comment by JeethJJ** [comment]
The mentioned issue is reproducible. 

https://github.com/user-attachments/assets/19cf1fe9-a008-4bc2-a2ff-7269037fd7c9

Although, the issue is only reproducible for sub org scenarios. I attempted the same case with 3 instances of IS as the app is available in IS1 (port 9443) and the user is in IS3 (port 9445). IS2 is in port 9444 and all are connected through OICD connections where IS3 is a federated idp for IS2 and IS2 is a federated idp for IS1. 

Eventhough through there is no OPBS cookie in OIDC logout request, and the logic was through handleMissingSessionStateGracefully method, the second login attempt was prompted for username & password as in the below recording. 

https://github.com/user-attachments/assets/a3ee1e10-da7b-4f12-a045-5b97635f2c1a

As per the offline chat with @SujanSanjula96, even if only present authenticator is the OIDC IDPs, the issue is not reproducible for multiple IS instances. 

https://github.com/user-attachments/assets/c28918c3-d1e3-4ebd-9e12-0c1411b015c9

Hence, this case should be specifically handled for Organization SSO authenticator,

**Comment by SujanSanjula96** [comment]
@JeethJJ Even though the login page is prompted, it should be due to an SSO issue . But session can be retrieved from the API call. Also, if we plug a different tenant in IS3 as the IDP, we can reproduce the issue with SSO as well.

**Comment by JeethJJ** [comment]
The issue is reproducible only if the attempted tenants/hostnames are the same. As mentioned in the docs [1] the hostnames has to be changed to configure OIDC federated IdP-initiated logout. With this configurations done right, the issue is not reproducible. 

Below is the expected behaviour.

https://github.com/user-attachments/assets/fe5963a4-b045-4acc-ae06-542d712a14bb

Further, this could be due to the cookies being overridden since they have the same hostname and tenant domains. With different tenant domains, the logout request for the sub-org is sent as expected (Image 1) and the cookies are sent with (Image 2).

Image 1
> <img width="1512" alt="Image" src="https://github.com/user-attachments/assets/defb183b-6067-4715-a1db-eabae9168287" />


Image 2
> <img width="1512" alt="Image" src="https://github.com/user-attachments/assets/0156b32a-f43d-4132-be0e-05e89a9cb769" />

As mentioned in the previous comments, the handleMissingSessionStateGracefully [2] method is introduced to handle error situations of cookie not existing gracefully by redirecting to the post_logout_redirect_uri [3].  

With this configuration disabled, the following behaviour will be observed if the OPBS cookie is not present in the logout request.

https://github.com/user-attachments/assets/0bd76c0b-1bcb-4e32-b2ec-b861954d4daf

The handleMissingSessionStateGracefully method was introduced only to handle this error state. It was not intended to handle logout when the cookie was not present. It was only to handle the error scenario by redirecting to the post_logout_redirect_uri.


[1] https://is.docs.wso2.com/en/6.0.0/guides/login/oidc-federated-idp-initiated-logout/#!:~:text=Since%20there%20can%20be%20issues%20with%20cookies%20when%20the%20same%20hostname%20is%20configured%20for%20both%20WSO2%20identity%20servers%20(primary%20and%20federated)%2C%20you%20need%20to%20configure%20different%20hostnames%20for%20both%20servers.

[2] https://github.com/wso2-extensions/identity-inbound-auth-oauth/blob/0c8ed64bac23ddf5d29b3af3cd411eda1683c16a/components/org.wso2.carbon.identity.oidc.session/src/main/java/org/wso2/carbon/identity/oidc/session/servlet/OIDCLogoutServlet.java#L173

[3] https://github.com/wso2/carbon-identity-framework/blob/38a5844e99db54b79b6c7844317f8c9a313ad1cc/features/identity-core/org.wso2.carbon.identity.core.server.feature/resources/identity.xml.j2#L1148
